pytorch_pfn_extras/training/extensions/best_value.py

```diff
@@ -0,0 +1,145 @@
+from typing import Any, Callable, Dict, Optional, TYPE_CHECKING
+
+from pytorch_pfn_extras import reporting
+from pytorch_pfn_extras.training import extension
+from pytorch_pfn_extras.training import trigger as trigger_module
+from pytorch_pfn_extras.training._manager_protocol import ExtensionsManagerProtocol
+
+
+if TYPE_CHECKING:
+    from pytorch_pfn_extras.training._trigger_util import TriggerLike
+
+
+class BestValue(extension.Extension):
+
+    """Extension traces the best value of a specific key in the observation.
+
+    Args:
+        key (str): Key of value.
+        compare (callable): Compare function which takes current best value and
+            new value and returns whether new value is better than current
+            best.
+        trigger: Trigger that decides the comparison interval between current
+            best value and new value. This must be a tuple in the form of
+            ``<int>, 'epoch'`` or ``<int>, 'iteration'`` which is passed to
+            :class:`~pytorch_pfn_extras.triggers.IntervalTrigger`.
+    """
+
+    default_name = 'best_value'
+
+    def __init__(
+            self,
+            key: str,
+            compare: Callable[[float, float], bool],
+            trigger: 'TriggerLike' = (1, 'epoch'),
+    ) -> None:
+        self._key = key
+        self._best_value: Optional[float] = None
+        self._best_epoch: Optional[int] = None
+        self._best_it: Optional[int] = None
+        self._interval_trigger = trigger_module.get_trigger(trigger)
+        self._init_summary()
+        self._compare = compare
+
+    def __call__(self, manager: ExtensionsManagerProtocol) -> None:
+        observation = manager.observation
+        summary = self._summary
+        key = self._key
+        if key in observation:
+            summary.add({key: observation[key]})
+
+        if not self._interval_trigger(manager):
+            return
+
+        stats = summary.compute_mean()
+        if key not in stats:
+            raise RuntimeError('Key "{}" not found in the observation '
+                               '(current available keys are {})'
+                               .format(key, list(stats.keys())))
+
+        value = float(stats[key])  # copy to CPU
```

Comment:
Can this copy be avoided? :)

Reply:
Fixed by re-using the `BestValueTrigger` internally!

Note: The small change applied in the BestValueTrigger [here](https://github.com/pfnet/pytorch-pfn-extras/pull/518/files#diff-5881bfe2983a95132b388de54eb31d5108a5bad3511c6e39138afcbaccdf331e) which is brought from my original code doesn't actually change the behavior, but just changes (enriches) the message in exceptions.

Reply:
`BestValue` syncs the cpu & gpu so if you call it multiple times inside an epoch, performance is likely going to suffer a lot.

Would it be possible to just do `torch.max(prev_value, new_value)` and do not use this inside any `if`?
I am not sure if this is possible, but what this aims is to have the best value at the end of the epoch calculated in the device without interrupting the stream.

Thanks!

Reply:
Aha, I see... I misunderstood that you meant *copy* by copying-and-pasting the core code from `BestValueTrigger`... I think that DRY would still be good so maybe I don't need to revert it, though.

BTW, the performance concern is not what is introduced by this PR (`BestValueTrigger` already has it), so may I leave it out of scope in this PR?
I'm also wondering a few points on introducing the suggested way, not sure they get the point though:
- In case the best value trigger object is loaded from a state dict, its internals are on CPU (`float`) at that moment, so comparison using `torch.{min,max}` would make things a little complicated to handle type/device inconsistency.
- Interfaces to get the best epoch/iters are sometimes useful; in case we are interested in not only the best values but also the actual model file if we save models e.g., every epochs. I guess `torch.max` approach cannot simply achieve this.

Reply:
no worries!

Reply:
FYI If you want to access the model parameters from the extension you need to set
`needs_model_state=True`

Reply:
Thanks for the comment as well as reviewing!
I wasn't aware of `needs_model_state=True` option. "corresponding model file" meant a little more stupid idea; one can easily get the filename of the best model file by `"model_epoch_{best_value_ext.best_epoch}"`